hv: Add 9P virtio peripheral

povik opened the pull request:
This allows one to access host m1n1 sources from within the system running in hypervisor. Once you arrange for the virtio peripheral both in hypervisor code (`self.attach_virtio(Virtio9PTransport(), 0x2_2000_0000, 382)`) and in the devicetree, you run

    mount -t 9p -o transp=virtio m1n1 /mnt

to mount the sources on guest. One use case of it is using m1n1 regmaps over kernel's internal regmap structures (as alluded to by the extra commit).

Performance is not great since the hypervisor stops to handle each 9P call but it's not impractical.

Direction I think this should be expanded in:

 * Allocate the IRQs/register range automatically
 * Insert the peripheral into ADT
 * Have guest m1n1 transfer it from ADT to FDT

Let me know if you think this is fine for inclusion.

marcan commented:
Very cute. I like it :)

I was going to say I'm not sure about putting this in the ADT, but on further thought there's no reason not to do that (as long as the mutation happens in Python, since we already have code for that there). So sure, why not!

marcan commented:
Just sort of direction-wise: I'm fine with anything in m1n1/the hypervisor as long as it helps with development. I don't think it should become a general-purpose hypervisor or veer off towards production usage, but sharing files with the host is *very* obviously useful for a ton of reasons (e.g. you can avoid building initramfses all the time for development, develop kernel modules and load/reload them from the guest without a full reboot, etc.). I've been doing this over sshfs in the past (host to guest, which sucks a bit), but the other way around with 9p should be nicer for a bunch of workflows.

povik commented:
Okay, I implemented the hypervisor->ADT->FDT path to pass information on the existence of the virtio device, and added allocation of IRQ/register range. The allocation looks for numbers/ranges that are free per ADT -- for MMIO that should be foolproof, and I guess the AIC isn't to any degree shared with the coprocessors, so for IRQs that should be okay too.

So, about usage: One passes a path and a tag to the hypervisor, so for example

    ./proxyclient/tools/run_guest.py \
        -v ~/repos/m1n1:m1n1 \
        -v ~/music:tunes \
        m1n1-payload.macho

to export one tree under tag `m1n1` and another under `tunes`. To then mount the tags on guest:

    mount -t 9p -o transp=virtio tunes /mnt
    mount -t 9p -o transp=virtio m1n1 /mnt2

You need a `u9fs` program installed on the host to do the 9P speaking, and these options on the guest kernel (let me know anybody if some are missing):

    CONFIG_NET_9P
    CONFIG_NET_9P_VIRTIO
    CONFIG_VIRTIO
    CONFIG_VIRTIO_MMIO
    CONFIG_9P_FS

povik commented on src/kboot.c:
```diff
@@ -1114,6 +1182,10 @@ int kboot_prepare_dt(void *fdt)
         return -1;
     if (dt_disable_missing_devs("i2c", "i2c@", 8))
         return -1;
+#ifndef RELEASE
```
Not sure if we want !RELEASE or some other gate here

povik commented:
Rebased

povik commented:
Rebased

povik commented:
Ping @marcan

marcan commented:
Looks good, can you fix the conflict so I can merge?

povik commented:
Sure, thanks for the review!

povik commented:
Rebased (and removed a stray definition in `uartproxy.h`)
